kolibri/core/assets/src/objectSpecs.js

```diff
@@ -0,0 +1,193 @@
+  }
+
+  // should only have a validator or a spec, not both
+  if (options.validator && options.spec) {
+    return _fail('Should either have a validator or a sub-spec', dataKey, data);
+  }
+
+  // validation function
+  if (hasData && options.validator && !options.validator(data)) {
+    return _fail('Validator function failed', dataKey, data);
+  }
+
+  // object sub-spec
+  if (hasData && options.spec) {
+    if (!isObject(data)) {
+      return _fail('Only objects can have sub-specs', dataKey, data);
+    }
+    if (!validateObject(data, options.spec)) {
+      return _fail('Validator sub-spec failed', dataKey, data);
+    }
+  }
+
+  // Check types
+  const KNOWN_TYPES = [Array, Boolean, Date, Function, Object, Number, String, Symbol];
+  if (isUndefined(options.type)) {
+    return _fail('No type information provided', dataKey, data);
+  } else if (!KNOWN_TYPES.includes(options.type)) {
+    return _fail(`Type '${options.type}' not currently handled`, dataKey);
+  } else if (hasData) {
+    if (options.type === Array && !isArray(data)) {
+      return _fail('Expected Array', dataKey, data);
+    } else if (options.type === Boolean && !isBoolean(data)) {
+      return _fail('Expected Boolean', dataKey, data);
+    } else if (options.type === Date && !isDate(data)) {
+      return _fail('Expected Date', dataKey, data);
+    } else if (options.type === Function && !isFunction(data)) {
+      return _fail('Expected Function', dataKey, data);
+    } else if (options.type === Object && !isObject(data)) {
+      return _fail('Expected Object', dataKey, data);
+    } else if (options.type === Number && !isNumber(data)) {
+      return _fail('Expected Number', dataKey, data);
+    } else if (options.type === String && !isString(data)) {
+      return _fail('Expected String', dataKey, data);
+    } else if (options.type === Symbol && !isSymbol(data)) {
+      return _fail('Expected Symbol', dataKey, data);
+    }
+  }
+
+  // ensure spec has a default when not required and not vice-versa
+  if (!options.required && isUndefined(options.default)) {
+    return _fail('Must be either required or have a default', dataKey, data);
+  } else if (options.required && !isUndefined(options.default)) {
+    return _fail('Cannot be required and have a default', dataKey, data);
+  }
+
+  // objects and arrays must use a generator function for their default value
+  if (
+    options.default &&
+    (options.type === Array || options.type === Object) &&
+    !isFunction(options.default)
+  ) {
+    return _fail('Need a function to return array and object default values', dataKey, data);
+  }
+
+  return true;
+}
+
+/*
+ * Given a JS object and a spec object as defined above, return `true` if the object
+ * conforms to the spec and `false` otherwise
+ */
+export function validateObject(object, spec) {
+  // skip validation in production
+  if (process.env.NODE_ENV === 'production') {
+    return true;
+  }
```

Comment:
Might be good to give a warning in this case otherwise you'll never know your object is failing the spec outside of production.